dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/MetadataVersionWrapper.java

```diff
@@ -0,0 +1,171 @@
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
+import org.hisp.dhis.common.DxfNamespaces;
+import org.hisp.dhis.keyjsonvalue.KeyJsonValue;
+import org.hisp.dhis.keyjsonvalue.KeyJsonValueService;
+import org.hisp.dhis.metadata.version.MetadataVersion;
+import org.hisp.dhis.metadata.version.MetadataVersionService;
+import org.hisp.dhis.render.RenderService;
+import org.hisp.dhis.system.startup.AbstractStartupRoutine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * @author David Katuscak
+ */
+public class MetadataVersionWrapper extends AbstractStartupRoutine
+{
+    private static final Logger log = LoggerFactory.getLogger( MetadataVersionWrapper.class );
+
+    private static final String METADATASTORE_NAMESPACE = "METADATASTORE";
+    private static final String WRAPPED_METADATA_PREFIX = "{\"metadata\":";
+
+    private final KeyJsonValueService keyJsonValueService;
+    private final MetadataVersionService metadataVersionService;
+    private final RenderService renderService;
+
+    public MetadataVersionWrapper( KeyJsonValueService keyJsonValueService,
+        MetadataVersionService metadataVersionService, RenderService renderService )
+    {
+        this.keyJsonValueService = keyJsonValueService;
+        this.metadataVersionService = metadataVersionService;
+        this.renderService = renderService;
+    }
+
+    @Override
+    public void execute() throws Exception
+    {
+        List<MetadataVersion> metadataVersions = metadataVersionService.getAllVersions();
+
+        for ( MetadataVersion metadataVersion : metadataVersions )
+        {
+            KeyJsonValue keyJsonValue = keyJsonValueService.getKeyJsonValue( METADATASTORE_NAMESPACE,
+                metadataVersion.getName() );
+
+            String value = keyJsonValue.getValue();
+            if ( value != null && !value.substring( 0, WRAPPED_METADATA_PREFIX.length() ).equals( WRAPPED_METADATA_PREFIX ) )
+            {
+                MetadataWrapper metadataWrapper = new MetadataWrapper( value );
+
+                //MetadataWrapper is used to avoid Metadata keys reordering by jsonb (jsonb does not preserve keys order)
+                keyJsonValue.setValue( renderService.toJsonAsString( metadataWrapper ) );
+                keyJsonValueService.updateKeyJsonValue( keyJsonValue );
+
+                metadataVersion.setHashCode( getHashCode( keyJsonValue.getValue() ) );
+
+                metadataVersionService.updateVersion( metadataVersion );
+            }
+        }
+    }
+
+    private static String getHashCode( String value ) throws NoSuchAlgorithmException
+    {
+        byte[] bytesOfMessage = value.getBytes( StandardCharsets.UTF_8 );
```

Comment:
This could be made public, or moved to utility class, and then add a test for it.

Reply:
It already is. I just copied it here as I couldn't reuse that class because I would introduce a circular dependency. Therefore, it is made private as it is aimed really only for use in this class. Otherwise, HashCodeGenerator.getHashCode() will be reused.